src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,577 @@
+        return of(
+                defaultInitializer(),
+                integrator,
+                Gatherers.Value.DEFAULT.statelessCombiner,
+                defaultFinisher()
+        );
+    }
+
+    /**
+     * Returns a new, parallelizable, and stateless {@code Gatherer} described
+     * by the given {@code integrator} and {@code finisher}.
+     *
+     * @param integrator the integrator function for the new gatherer
+     * @param finisher the finisher function for the new gatherer
+     * @param <T> the type of input elements for the new gatherer
+     * @param <R> the type of results for the new gatherer
+     * @throws NullPointerException if any argument is null
+     * @return the new {@code Gatherer}
+     */
+    static <T, R> Gatherer<T, Void, R> of(
+            Integrator<Void, T, R> integrator,
+            BiConsumer<Void, Downstream<? super R>> finisher) {
+        return of(
+                defaultInitializer(),
+                integrator,
+                Gatherers.Value.DEFAULT.statelessCombiner,
+                finisher
+        );
+    }
+
+    /**
+     * Returns a new, parallelizable, {@code Gatherer} described by the given
+     * {@code initializer}, {@code integrator}, {@code combiner} and
+     * {@code finisher}.
+     *
+     * @param initializer the supplier function for the new gatherer
+     * @param integrator the integrator function for the new gatherer
+     * @param combiner the combiner function for the new gatherer
+     * @param finisher the finisher function for the new gatherer
+     * @param <T> the type of input elements for the new gatherer
+     * @param <A> the type of initializer for the new gatherer
+     * @param <R> the type of results for the new gatherer
+     * @throws NullPointerException if any argument is null
+     * @return the new {@code Gatherer}
+     */
+    static <T, A, R> Gatherer<T, A, R> of(
+            Supplier<A> initializer,
+            Integrator<A, T, R> integrator,
+            BinaryOperator<A> combiner,
+            BiConsumer<A, Downstream<? super R>> finisher) {
+        return new Gatherers.GathererImpl<>(
+                Objects.requireNonNull(initializer),
+                Objects.requireNonNull(integrator),
+                Objects.requireNonNull(combiner),
+                Objects.requireNonNull(finisher)
+        );
+    }
+
+    /**
+     * A Downstream object is the next stage in a pipeline of operations,
+     * to which elements can be sent.
+     * @param <T> the type of elements this downstream accepts
+     * @since 22
+     */
+    @FunctionalInterface
+    @PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+    interface Downstream<T> {
+
+        /**
+         * Pushes, if possible, the provided element to the next stage in the
+         * pipeline.
+         *
+         * <p>If this method returns {@code false} then the next stage does
+         * not accept any more elements.
```

Comment:
So once push returns false then any further attempts to push to this downstream will also return false.

Reply:
Yeah, perhaps should read something like:

`If this method returns {@code false} then no further elements need to be pushed, and subsequent pushes will also return {@code false}.` ? (And make it an implSpec?)